core/src/main/java/io/questdb/cutlass/http/WaitProcessor.java

```diff
@@ -0,0 +1,211 @@
+    private final Sequence outSubSequence = new MCSequence(retryQueueLength);
+    private final MillisecondClock clock;
+    private final long maxWaitCapMs;
+    private final double exponentialWaitMultiplier;
+
+    public WaitProcessor(WaitProcessorConfiguration configuration) {
+        this.clock = configuration.getClock();
+        this.maxWaitCapMs = configuration.getMaxWaitCapMs();
+        this.exponentialWaitMultiplier = configuration.getExponentialWaitMultiplier();
+
+        inPubSequence.then(inSubSequence).then(inPubSequence);
+        outPubSequence.then(outSubSequence).then(outPubSequence);
+    }
+
+    @Override
+    protected boolean runSerially() {
+        return processInQueue() || sendToOutQueue();
+    }
+
+    @Override
+    // This supposed to run in http execution thread / job
+    public void reschedule(Retry retry) {
+        reschedule(retry, 0, 0);
+    }
+
+    private void reschedule(Retry retry, int attempt, long waitStartMs) {
+        long now = clock.getTicks();
+        retry.getAttemptDetails().attempt = attempt;
+        retry.getAttemptDetails().lastRunTimestamp = now;
+        retry.getAttemptDetails().waitStartTimestamp = attempt == 0 ? now : waitStartMs;
+
+        while (true) {
+            long cursor = inPubSequence.next();
+            // -2 = there was a contest for queue index and this thread has lost
+            if (cursor < -1) {
+                continue;
+            }
+
+            // -1 = queue is empty. It means there are already too many retries waiting
+            // Send error to client.
+            if (cursor < 0) {
+                throw EntryUnavailableException.INSTANCE;
+            }
+
+            try {
+                inQueue.get(cursor).retry = retry;;
+                return;
+            } finally {
+                inPubSequence.done(cursor);
+            }
+        }
+    }
+
+    // This hijacks http execution thread / job and runs retries in it.
+    public boolean runReruns(HttpRequestProcessorSelector selector) {
+        boolean useful = false;
+
+        while (true) {
+            RetryHolder retryHolder = getNextRerun();
+            if (retryHolder != null) {
+                useful = true;
+                if (!retryHolder.retry.tryRerun(selector)) {
+                    // Need more attempts
+                    reschedule(retryHolder.retry, retryHolder.retry.getAttemptDetails().attempt + 1, retryHolder.retry.getAttemptDetails().waitStartTimestamp);
+                }
+            } else {
+                return useful;
+            }
+        }
+    }
+
+    private @Nullable RetryHolder getNextRerun() {
+        long cursor = outSubSequence.next();
+        // -2 = there was a contest for queue index and this thread has lost
+        if (cursor < 0) {
+            return null;
+        }
+
+        try {
+            return outQueue.get(cursor);
+        } finally {
+            outSubSequence.done(cursor);
```

Comment:
This is a race condition. done() releases slot, which can be written to right after done() and right before RetryHolder is retried. This can lead to unpredictable results.

RetryHolder needs to be copied off before releasing queue slot